File: backend/discussion_cache.py
"""backend/discussion_cache.py — SQLite cache for GitHub Discussion reads.

Cuts GraphQL traffic by serving repeated reads from a local cache with a 300s TTL.
Cache lives at $AUTONOMOUS_TEAM_STATE_DIR/discussion_cache.db (never inside the repo).

Public API:
    get_body(number, fresh=False) -> str   cached body, or a fresh fetch when fresh=True;
                                            "" on total failure
    get(number, fresh=False) -> dict       full record: body, title, labels, updated_at
    list_open() -> list[dict]              ONE batched GraphQL call for all open discussions
    invalidate(number)                     force next read to re-fetch

CLI (called by shell scripts):
    python3 backend/discussion_cache.py get-body <N> [--fresh]
    python3 backend/discussion_cache.py list-open
    python3 backend/discussion_cache.py invalidate <N>

`fresh=True` bypasses the TTL check and re-fetches unconditionally. It exists for the
handful of correctness-critical reads (the spawn-gate's SPEC_READY check) where a stale
body can misdirect the caller; the bulk/context-assembly readers keep the plain TTL-cached
path on purpose — that's what the cache is for.

On GraphQL failure the cache returns stale data with a stderr warning rather than
raising — the calling script sees a non-empty result and can continue. This holds for
fresh reads too: a fresh read that hard-fails on a transient GraphQL blip would trade an
intermittent staleness bug for an outage, so it falls back to the same stale-cache
behaviour. The CLI's `get-body --fresh` distinguishes this case via exit code 3 (see
below) so callers that must not silently trust a stale fallback can react to it.

CLI exit codes for `get-body`:
    0   body printed to stdout, and (for --fresh) it is a live/current read
    1   nothing available at all — stdout is empty
    3   --fresh was requested but the live fetch failed; stdout holds a *stale*
        cached fallback (also flagged via a stderr warning)
"""

from __future__ import annotations

import json
import sqlite3
import subprocess
import sys
import time
import warnings
from pathlib import Path
from typing import Optional

# ---------------------------------------------------------------------------
# Paths
# ---------------------------------------------------------------------------

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from backend import state_paths as _state_paths  # noqa: E402
from backend.state_paths import ensure_state_dir  # noqa: E402
from backend._repo import REPO as _REPO, REPO_OWNER as _REPO_OWNER, REPO_NAME as _REPO_NAME  # noqa: E402

_TTL: int = 300  # seconds


def _db_path() -> Path:
    # Resolved at call time, not import time — see D#1810.
    return _state_paths.STATE_DIR / "discussion_cache.db"

# ---------------------------------------------------------------------------
# Schema bootstrap
# ---------------------------------------------------------------------------

_DDL = """
CREATE TABLE IF NOT EXISTS discussion_cache (
    number     INTEGER PRIMARY KEY,
    body       TEXT    NOT NULL DEFAULT '',
    title      TEXT    NOT NULL DEFAULT '',
    labels     TEXT    NOT NULL DEFAULT '[]',
    updated_at TEXT    NOT NULL DEFAULT '',
    cached_at  TEXT    NOT NULL DEFAULT ''
);
"""


def _conn() -> sqlite3.Connection:
    ensure_state_dir()
    con = sqlite3.connect(str(_db_path()))
    con.row_factory = sqlite3.Row
    con.execute(_DDL)
    con.commit()
    return con


# ---------------------------------------------------------------------------
# Stats counters (hit/miss) — stored in same db
# ---------------------------------------------------------------------------

_COUNTER_DDL = """
CREATE TABLE IF NOT EXISTS discussion_cache_stats (
    key   TEXT PRIMARY KEY,
    value INTEGER NOT NULL DEFAULT 0
);
"""


def _inc(key: str) -> None:
    """Increment a named counter (best-effort — never raises)."""
    try:
        with _conn() as con:
            con.execute(_COUNTER_DDL)
            con.execute(
                "INSERT INTO discussion_cache_stats(key, value) VALUES(?, 1) "
                "ON CONFLICT(key) DO UPDATE SET value = value + 1",
                (key,),
            )
    except Exception:  # noqa: BLE001
        pass


def get_stats() -> dict:
    """Return hit/miss/total counts and hit_ratio (float 0–1)."""
    try:
        with _conn() as con:
            con.execute(_COUNTER_DDL)
            rows = con.execute("SELECT key, value FROM discussion_cache_stats").fetchall()
        counts = {r["key"]: r["value"] for r in rows}
    except Exception:  # noqa: BLE001
        counts = {}
    hits = counts.get("hit", 0)
    misses = counts.get("miss", 0)
    total = hits + misses
    return {
        "hits": hits,
        "misses": misses,
        "total": total,
        "hit_ratio": round(hits / total, 4) if total else 0.0,
    }


# ---------------------------------------------------------------------------
# GraphQL helpers
# ---------------------------------------------------------------------------

def _gh_graphql(query: str) -> Optional[dict]:
    """Run a gh graphql call and return parsed JSON, or None on error."""
    try:
        result = subprocess.run(
            ["gh", "api", "graphql", "-f", f"query={query}"],
            capture_output=True,
            text=True,
            timeout=30,
        )
        if result.returncode != 0:
            warnings.warn(
                f"discussion_cache: gh graphql failed: {result.stderr.strip()}",
                stacklevel=2,
            )
            return None
        return json.loads(result.stdout)
    except Exception as exc:  # noqa: BLE001
        warnings.warn(f"discussion_cache: gh graphql error: {exc}", stacklevel=2)
        return None


def _fetch_one(number: int) -> Optional[dict]:
    """Fetch a single discussion from GitHub. Returns dict or None on failure."""
    query = (
        f'query {{ repository(owner:"{_REPO_OWNER}", name:"{_REPO_NAME}") {{'
        f' discussion(number: {number}) {{ title body updatedAt labels(first:10) {{ nodes {{ name }} }} }} }} }}'
    )
    data = _gh_graphql(query)
    if not data:
        return None
    try:
        disc = data["data"]["repository"]["discussion"]
        if disc is None:
            return None
        labels = [n["name"] for n in (disc.get("labels", {}).get("nodes") or [])]
        return {
            "number": number,
            "title": disc.get("title", ""),
            "body": disc.get("body", ""),
            "labels": labels,
            "updated_at": disc.get("updatedAt", ""),
        }
    except (KeyError, TypeError):
        return None


def _fetch_all_open() -> Optional[list[dict]]:
    """Fetch all open discussions in one batched GraphQL call."""
    query = (
        f'query {{ repository(owner:"{_REPO_OWNER}", name:"{_REPO_NAME}") {{'
        ' discussions(first:100, states:[OPEN]) { nodes { number title body updatedAt'
        ' labels(first:10) { nodes { name } } } } } }'
    )
    data = _gh_graphql(query)
    if not data:
        return None
    try:
        nodes = data["data"]["repository"]["discussions"]["nodes"]
        result = []
        for disc in nodes:
            labels = [n["name"] for n in (disc.get("labels", {}).get("nodes") or [])]
            result.append({
                "number": disc["number"],
                "title": disc.get("title", ""),
                "body": disc.get("body", ""),
                "labels": labels,
                "updated_at": disc.get("updatedAt", ""),
            })
        return result
    except (KeyError, TypeError):
        return None


# ---------------------------------------------------------------------------
# Cache helpers
# ---------------------------------------------------------------------------

def _now_iso() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


def _is_fresh(cached_at: str) -> bool:
    if not cached_at:
        return False
    try:
        t = time.mktime(time.strptime(cached_at, "%Y-%m-%dT%H:%M:%SZ"))
        return (time.time() - t) < _TTL
    except ValueError:
        return False


def _cache_row(con: sqlite3.Connection, record: dict) -> None:
    labels_json = json.dumps(record.get("labels", []))
    con.execute(
        "INSERT INTO discussion_cache(number, body, title, labels, updated_at, cached_at) "
        "VALUES(?,?,?,?,?,?) "
        "ON CONFLICT(number) DO UPDATE SET "
        "  body=excluded.body, title=excluded.title, labels=excluded.labels, "
        "  updated_at=excluded.updated_at, cached_at=excluded.cached_at",
        (
            record["number"],
            record.get("body", ""),
            record.get("title", ""),
            labels_json,
            record.get("updated_at", ""),
            _now_iso(),
        ),
    )


def _read_row(con: sqlite3.Connection, number: int) -> Optional[sqlite3.Row]:
    return con.execute(
        "SELECT * FROM discussion_cache WHERE number = ?", (number,)
    ).fetchone()


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

# Status values returned by _get_record: what kind of read produced the result.
#   "cached"        — served from a TTL-fresh cache row, no fetch attempted
#   "fetched"       — live GraphQL fetch succeeded (fresh=True or cache miss/stale)
#   "stale_fallback"— fetch failed, served a stale cached row instead
#   "empty"         — fetch failed and nothing was cached either

def _get_record(number: int, fresh: bool) -> tuple[Optional[dict], str]:
    """Shared read path for get_body/get. Returns (record_dict_or_None, status).

    record_dict has keys: number, title, body, labels, updated_at, cached_at.
    """
    with _conn() as con:
        row = _read_row(con, number)
        if row and not fresh and _is_fresh(row["cached_at"]):
            _inc("hit")
            return {
                "number": number,
                "title": row["title"],
                "body": row["body"],
                "labels": json.loads(row["labels"] or "[]"),
                "updated_at": row["updated_at"],
                "cached_at": row["cached_at"],
            }, "cached"

    # Cache miss, stale, or fresh=True forcing a live read — fetch.
    _inc("miss")
    record = _fetch_one(number)

    if record is None:
        # Graceful degradation: return stale value if available rather than failing
        # hard. This applies to fresh=True reads too — a transient GraphQL blip must
        # not turn an intermittent staleness bug into a spawn-gate outage.
        if row:
            sys.stderr.write(
                f"[discussion_cache] WARNING: GraphQL failed, returning stale body for #{number}\n"
            )
            return {
                "number": number,
                "title": row["title"],
                "body": row["body"],
                "labels": json.loads(row["labels"] or "[]"),
                "updated_at": row["updated_at"],
                "cached_at": row["cached_at"],
            }, "stale_fallback"
        return None, "empty"

    with _conn() as con:
        _cache_row(con, record)

    return {**record, "cached_at": _now_iso()}, "fetched"


def get_body(number: int, fresh: bool = False) -> str:
    """Return the discussion body.

    fresh=False (default): serve from cache when the TTL is still fresh — this is
        the bulk/context-assembly path the cache exists to optimise.
    fresh=True: bypass the TTL and re-fetch unconditionally. On GraphQL failure this
        still falls back to a stale cached value (with a stderr warning) rather than
        failing hard — use `get_body_status` if the caller needs to distinguish a
        genuinely fresh read from a stale fallback.

    Returns "" when nothing is available at all.
    """
    record, _status = _get_record(number, fresh)
    return record["body"] if record else ""


def get_body_status(number: int, fresh: bool = False) -> tuple[str, str]:
    """Like get_body, but also returns the read status.

    Returns (body, status) where status is one of "cached", "fetched",
    "stale_fallback", "empty". Callers that must not silently trust a stale
    fallback (e.g. the spawn gate) should branch on status rather than just body.
    """
    record, status = _get_record(number, fresh)
    return (record["body"] if record else ""), status


def get(number: int, fresh: bool = False) -> dict:
    """Return full cached record. Fetches if stale/missing, or unconditionally if fresh=True."""
    record, _status = _get_record(number, fresh)
    return record or {}


def list_open() -> list[dict]:
    """Fetch ALL open discussions in one GraphQL call, update cache, return list."""
    records = _fetch_all_open()
    if records is None:
        # Fallback: return whatever we have cached
        sys.stderr.write(
            "[discussion_cache] WARNING: GraphQL failed in list_open, returning stale cache\n"
        )
        with _conn() as con:
            rows = con.execute("SELECT * FROM discussion_cache").fetchall()
        return [
            {
                "number": r["number"],
                "title": r["title"],
                "body": r["body"],
                "labels": json.loads(r["labels"] or "[]"),
                "updated_at": r["updated_at"],
                "cached_at": r["cached_at"],
            }
            for r in rows
        ]

    with _conn() as con:
        for record in records:
            _cache_row(con, record)

    return [{**r, "cached_at": _now_iso()} for r in records]


def invalidate(number: int) -> None:
    """Clear cache entry so next read forces a fresh GraphQL fetch."""
    with _conn() as con:
        con.execute(
            "UPDATE discussion_cache SET cached_at = '' WHERE number = ?", (number,)
        )


# ---------------------------------------------------------------------------
# CLI entrypoint
# ---------------------------------------------------------------------------

def _usage() -> None:
    sys.stderr.write(
        "Usage:\n"
        "  python3 backend/discussion_cache.py get-body <N> [--fresh]\n"
        "  python3 backend/discussion_cache.py get <N> [--fresh]\n"
        "  python3 backend/discussion_cache.py list-open\n"
        "  python3 backend/discussion_cache.py invalidate <N>\n"
        "  python3 backend/discussion_cache.py stats\n"
        "\n"
        "get-body exit codes: 0=printed (live if --fresh), 1=nothing available,\n"
        "  3=--fresh requested but fetch failed; stdout holds a STALE fallback.\n"
    )


if __name__ == "__main__":
    args = sys.argv[1:]
    if not args:
        _usage()
        sys.exit(1)

    cmd = args[0]

    if cmd == "get-body":
        if len(args) < 2:
            sys.stderr.write("get-body requires a discussion number\n")
            sys.exit(1)
        fresh = "--fresh" in args[2:]
        body, status = get_body_status(int(args[1]), fresh=fresh)
        sys.stdout.write(body)
        if not body:
            sys.exit(1)
        if fresh and status == "stale_fallback":
            sys.exit(3)
        sys.exit(0)

    elif cmd == "get":
        if len(args) < 2:
            sys.stderr.write("get requires a discussion number\n")
            sys.exit(1)
        fresh = "--fresh" in args[2:]
        record = get(int(args[1]), fresh=fresh)
        print(json.dumps(record, indent=2))

    elif cmd == "list-open":
        records = list_open()
        print(json.dumps(records, indent=2))

    elif cmd == "invalidate":
        if len(args) < 2:
            sys.stderr.write("invalidate requires a discussion number\n")
            sys.exit(1)
        invalidate(int(args[1]))
        print(f"invalidated #{args[1]}")

    elif cmd == "stats":
        print(json.dumps(get_stats(), indent=2))

    else:
        sys.stderr.write(f"Unknown command: {cmd}\n")
        _usage()
        sys.exit(1)

File: ts-backend/src/spawn/discussion-cache.ts
/**
 * spawn/discussion-cache.ts — SQLite cache for GitHub Discussion reads.
 *
 * Faithful 1:1 port of backend/discussion_cache.py (401 LOC).
 *
 * Mirrors:
 *   get_body(number) → getBody(number): string
 *   get(number)      → getDiscussion(number): DiscussionRecord | {}
 *   list_open()      → listOpen(): DiscussionRecord[]
 *   invalidate(number) → invalidate(number): void
 *   get_stats()      → getStats(): CacheStats
 *
 * Cache lives at $AUTONOMOUS_TEAM_STATE_DIR/discussion_cache.db (never inside the repo).
 * TTL: 300 seconds. On GraphQL failure, stale data is returned with a stderr warning
 * rather than throwing — callers see a non-empty result and can continue.
 *
 * CLI (mirroring Python exactly):
 *   bun run src/spawn/discussion-cache.ts get-body <N>
 *   bun run src/spawn/discussion-cache.ts get <N>
 *   bun run src/spawn/discussion-cache.ts list-open
 *   bun run src/spawn/discussion-cache.ts invalidate <N>
 *   bun run src/spawn/discussion-cache.ts stats
 */

import { Database } from "bun:sqlite";
import { mkdirSync, existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { execFileSync } from "node:child_process";
import { stateDir as sharedStateDir } from "../config/state-paths.js";

// ---------------------------------------------------------------------------
// Paths — mirrors backend/state_paths.py resolution order
// ---------------------------------------------------------------------------

function stateDir(): string {
  return sharedStateDir();
}

export function dbPath(): string {
  return (
    process.env["DISCUSSION_CACHE_DB_PATH"] ??
    join(stateDir(), "discussion_cache.db")
  );
}

/** Mirrors ensure_state_dir() — idempotent. */
function ensureStateDir(): void {
  const dir = stateDir();
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true });
  }
  const blackboard = join(dir, "blackboard");
  if (!existsSync(blackboard)) {
    mkdirSync(blackboard, { recursive: true });
  }
}

// ---------------------------------------------------------------------------
// Repo resolution — mirrors backend/_repo.py resolution order
// ---------------------------------------------------------------------------

function loadRepo(): { owner: string; name: string } {
  const envRepo = process.env["AUTONOMOUS_TEAM_REPO"];
  if (envRepo) {
    const parts = envRepo.split("/", 2);
    return { owner: parts[0] ?? "", name: parts[1] ?? "" };
  }

  // State-dir project.json
  const stateProjectJson = join(stateDir(), "project.json");
  if (existsSync(stateProjectJson)) {
    try {
      const data = JSON.parse(readFileSync(stateProjectJson, "utf-8")) as Record<string, unknown>;
      const repo = data["repo"] as string | undefined;
      if (repo) {
        const parts = repo.split("/", 2);
        return { owner: parts[0] ?? "", name: parts[1] ?? "" };
      }
    } catch { /* ignore */ }
  }

  // Repo-root .autonomous-team/project.json (backwards compat)
  // This file lives at ts-backend/src/spawn/discussion-cache.ts
  // Repo root = 4 levels up
  const repoRoot =
    process.env["AF_REPO_ROOT"] ??
    join(new URL(import.meta.url).pathname, "..", "..", "..", "..");
  const repoProjectJson = join(repoRoot, ".autonomous-team", "project.json");
  if (existsSync(repoProjectJson)) {
    try {
      const data = JSON.parse(readFileSync(repoProjectJson, "utf-8")) as Record<string, unknown>;
      const repo = data["repo"] as string | undefined;
      if (repo) {
        const parts = repo.split("/", 2);
        return { owner: parts[0] ?? "", name: parts[1] ?? "" };
      }
    } catch { /* ignore */ }
  }

  return { owner: "fulcrumaxe", name: "fulcrumaxe" };
}

const _REPO = loadRepo();
const _REPO_OWNER = _REPO.owner;
const _REPO_NAME = _REPO.name;

// ---------------------------------------------------------------------------
// TTL
// ---------------------------------------------------------------------------

const _TTL_SECONDS = 300;

// ---------------------------------------------------------------------------
// Schema DDL — mirrors Python _DDL and _COUNTER_DDL exactly
// ---------------------------------------------------------------------------

const _DDL = `
CREATE TABLE IF NOT EXISTS discussion_cache (
    number     INTEGER PRIMARY KEY,
    body       TEXT    NOT NULL DEFAULT '',
    title      TEXT    NOT NULL DEFAULT '',
    labels     TEXT    NOT NULL DEFAULT '[]',
    updated_at TEXT    NOT NULL DEFAULT '',
    cached_at  TEXT    NOT NULL DEFAULT ''
);
`;

const _COUNTER_DDL = `
CREATE TABLE IF NOT EXISTS discussion_cache_stats (
    key   TEXT PRIMARY KEY,
    value INTEGER NOT NULL DEFAULT 0
);
`;

// ---------------------------------------------------------------------------
// DB connection factory — mirrors _conn()
// ---------------------------------------------------------------------------

export function openDb(): Database {
  ensureStateDir();
  const db = new Database(dbPath(), { create: true });
  db.run(_DDL);
  db.run("PRAGMA journal_mode=WAL");
  return db;
}

// ---------------------------------------------------------------------------
// Stats counters — mirrors _inc() + get_stats()
// ---------------------------------------------------------------------------

function inc(key: string): void {
  try {
    const db = openDb();
    db.run(_COUNTER_DDL);
    db.run(
      "INSERT INTO discussion_cache_stats(key, value) VALUES(?, 1) " +
        "ON CONFLICT(key) DO UPDATE SET value = value + 1",
      [key]
    );
    db.close();
  } catch {
    /* best-effort — never raises, mirrors Python pass */
  }
}

export interface CacheStats {
  hits: number;
  misses: number;
  total: number;
  hit_ratio: number;
}

export function getStats(): CacheStats {
  try {
    const db = openDb();
    db.run(_COUNTER_DDL);
    const rows = db
      .query<{ key: string; value: number }, []>(
        "SELECT key, value FROM discussion_cache_stats"
      )
      .all();
    db.close();
    const counts: Record<string, number> = {};
    for (const r of rows) {
      counts[r.key] = r.value;
    }
    const hits = counts["hit"] ?? 0;
    const misses = counts["miss"] ?? 0;
    const total = hits + misses;
    return {
      hits,
      misses,
      total,
      hit_ratio: total > 0 ? Math.round((hits / total) * 10000) / 10000 : 0.0,
    };
  } catch {
    return { hits: 0, misses: 0, total: 0, hit_ratio: 0.0 };
  }
}

// ---------------------------------------------------------------------------
// Time helpers — mirrors _now_iso() and _is_fresh()
// ---------------------------------------------------------------------------

export function nowIso(): string {
  // Mirrors Python: time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
  // No milliseconds — match Python format exactly
  const d = new Date();
  const pad = (n: number): string => String(n).padStart(2, "0");
  return (
    `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}` +
    `T${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}Z`
  );
}

export function isFresh(cachedAt: string): boolean {
  if (!cachedAt) return false;
  try {
    const t = new Date(cachedAt).getTime();
    if (isNaN(t)) return false;
    return (Date.now() - t) / 1000 < _TTL_SECONDS;
  } catch {
    return false;
  }
}

// ---------------------------------------------------------------------------
// GraphQL helpers — mirrors _gh_graphql, _fetch_one, _fetch_all_open
// ---------------------------------------------------------------------------

function ghGraphql(query: string): Record<string, unknown> | null {
  try {
    // Use execFileSync (no shell) to mirror Python's subprocess.run(["gh","api","graphql","-f",f"query={query}"], ...)
    // This prevents shell metacharacter injection via query string or repo owner/name.
    const stdout = execFileSync("gh", ["api", "graphql", "-f", `query=${query}`], {
      timeout: 30_000,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    });
    return JSON.parse(stdout) as Record<string, unknown>;
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    process.stderr.write(`discussion_cache: gh graphql error: ${msg}\n`);
    return null;
  }
}

export interface DiscussionRecord {
  number: number;
  title: string;
  body: string;
  labels: string[];
  updated_at: string;
  cached_at?: string;
}

export function fetchOne(number: number): DiscussionRecord | null {
  const query =
    `query { repository(owner:"${_REPO_OWNER}", name:"${_REPO_NAME}") {` +
    ` discussion(number: ${number}) { title body updatedAt labels(first:10) { nodes { name } } } } }`;
  const data = ghGraphql(query);
  if (!data) return null;
  try {
    type GqlResp = {
      data: {
        repository: {
          discussion: Record<string, unknown> | null;
        };
      };
    };
    const disc = (data as GqlResp).data.repository.discussion;
    if (disc === null || disc === undefined) return null;
    const labelsObj = disc["labels"] as
      | { nodes: { name: string }[] }
      | undefined;
    const labelsNodes = labelsObj?.nodes ?? [];
    const labels = labelsNodes.map((n) => n.name);
    return {
      number,
      title: (disc["title"] as string | undefined) ?? "",
      body: (disc["body"] as string | undefined) ?? "",
      labels,
      updated_at: (disc["updatedAt"] as string | undefined) ?? "",
    };
  } catch {
    return null;
  }
}

export function fetchAllOpen(): DiscussionRecord[] | null {
  const query =
    `query { repository(owner:"${_REPO_OWNER}", name:"${_REPO_NAME}") {` +
    " discussions(first:100, states:[OPEN]) { nodes { number title body updatedAt" +
    " labels(first:10) { nodes { name } } } } } }";
  const data = ghGraphql(query);
  if (!data) return null;
  try {
    type GqlResp = {
      data: {
        repository: {
          discussions: { nodes: Array<Record<string, unknown>> };
        };
      };
    };
    const nodes = (data as GqlResp).data.repository.discussions.nodes;
    return nodes.map((disc) => {
      const labelsObj = disc["labels"] as
        | { nodes: { name: string }[] }
        | undefined;
      const labelsNodes = labelsObj?.nodes ?? [];
      const labels = labelsNodes.map((n) => n.name);
      return {
        number: disc["number"] as number,
        title: (disc["title"] as string | undefined) ?? "",
        body: (disc["body"] as string | undefined) ?? "",
        labels,
        updated_at: (disc["updatedAt"] as string | undefined) ?? "",
      };
    });
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Cache row helpers — mirrors _cache_row and _read_row
// ---------------------------------------------------------------------------

interface CacheRow {
  number: number;
  body: string;
  title: string;
  labels: string;
  updated_at: string;
  cached_at: string;
}

export function cacheRow(db: Database, record: DiscussionRecord): void {
  const labelsJson = JSON.stringify(record.labels ?? []);
  db.run(
    "INSERT INTO discussion_cache(number, body, title, labels, updated_at, cached_at) " +
      "VALUES(?,?,?,?,?,?) " +
      "ON CONFLICT(number) DO UPDATE SET " +
      "  body=excluded.body, title=excluded.title, labels=excluded.labels, " +
      "  updated_at=excluded.updated_at, cached_at=excluded.cached_at",
    [
      record.number,
      record.body ?? "",
      record.title ?? "",
      labelsJson,
      record.updated_at ?? "",
      nowIso(),
    ]
  );
}

export function readRow(db: Database, number: number): CacheRow | null {
  const row = db
    .query<CacheRow, [number]>(
      "SELECT * FROM discussion_cache WHERE number = ?"
    )
    .get(number);
  return row ?? null;
}

// ---------------------------------------------------------------------------
// Public API — mirrors Python exactly
// ---------------------------------------------------------------------------

/**
 * Return the discussion body, using cache if TTL fresh.
 *
 * On GraphQL failure: returns stale cached value (if any) with a stderr warning.
 * Returns "" when nothing is available.
 * Mirrors: get_body(number) → str
 */
export function getBody(number: number): string {
  const db = openDb();
  const row = readRow(db, number);
  if (row && isFresh(row.cached_at)) {
    db.close();
    inc("hit");
    return row.body;
  }
  db.close();

  // Cache miss or stale — fetch
  inc("miss");
  const record = fetchOne(number);

  if (record === null) {
    // Graceful degradation: return stale value if available
    if (row) {
      process.stderr.write(
        `[discussion_cache] WARNING: GraphQL failed, returning stale body for #${number}\n`
      );
      return row.body;
    }
    return "";
  }

  const db2 = openDb();
  cacheRow(db2, record);
  db2.close();

  return record.body;
}

/**
 * Return full cached record. Fetches if stale/missing.
 * Mirrors: get(number) → dict
 */
export function getDiscussion(
  number: number
): DiscussionRecord | Record<string, never> {
  const db = openDb();
  const row = readRow(db, number);
  if (row && isFresh(row.cached_at)) {
    db.close();
    inc("hit");
    return {
      number,
      title: row.title,
      body: row.body,
      labels: JSON.parse(row.labels || "[]") as string[],
      updated_at: row.updated_at,
      cached_at: row.cached_at,
    };
  }
  db.close();

  inc("miss");
  const record = fetchOne(number);

  if (record === null) {
    if (row) {
      process.stderr.write(
        `[discussion_cache] WARNING: GraphQL failed, returning stale record for #${number}\n`
      );
      return {
        number,
        title: row.title,
        body: row.body,
        labels: JSON.parse(row.labels || "[]") as string[],
        updated_at: row.updated_at,
        cached_at: row.cached_at,
      };
    }
    return {};
  }

  const db2 = openDb();
  cacheRow(db2, record);
  db2.close();

  return { ...record, cached_at: nowIso() };
}

/**
 * Fetch ALL open discussions in one GraphQL call, update cache, return list.
 * Mirrors: list_open() → list[dict]
 */
export function listOpen(): DiscussionRecord[] {
  const records = fetchAllOpen();

  if (records === null) {
    // Fallback: return whatever we have cached
    process.stderr.write(
      "[discussion_cache] WARNING: GraphQL failed in list_open, returning stale cache\n"
    );
    const db = openDb();
    const rows = db
      .query<CacheRow, []>("SELECT * FROM discussion_cache")
      .all();
    db.close();
    return rows.map((r) => ({
      number: r.number,
      title: r.title,
      body: r.body,
      labels: JSON.parse(r.labels || "[]") as string[],
      updated_at: r.updated_at,
      cached_at: r.cached_at,
    }));
  }

  const db = openDb();
  for (const record of records) {
    cacheRow(db, record);
  }
  db.close();

  return records.map((r) => ({ ...r, cached_at: nowIso() }));
}

/**
 * Clear cache entry so next read forces a fresh GraphQL fetch.
 * Mirrors: invalidate(number) → None
 */
export function invalidate(number: number): void {
  const db = openDb();
  db.run("UPDATE discussion_cache SET cached_at = '' WHERE number = ?", [
    number,
  ]);
  db.close();
}

// ---------------------------------------------------------------------------
// CLI entrypoint — mirrors Python __main__ block exactly
// ---------------------------------------------------------------------------

function usage(): void {
  process.stderr.write(
    "Usage:\n" +
      "  bun run src/spawn/discussion-cache.ts get-body <N>\n" +
      "  bun run src/spawn/discussion-cache.ts get <N>\n" +
      "  bun run src/spawn/discussion-cache.ts list-open\n" +
      "  bun run src/spawn/discussion-cache.ts invalidate <N>\n" +
      "  bun run src/spawn/discussion-cache.ts stats\n"
  );
}

if (import.meta.main) {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    usage();
    process.exit(1);
  }

  const cmd = args[0];

  if (cmd === "get-body") {
    if (args.length < 2) {
      process.stderr.write("get-body requires a discussion number\n");
      process.exit(1);
    }
    const body = getBody(parseInt(args[1]!, 10));
    process.stdout.write(body);
    // exit 1 if nothing returned so callers can detect missing — mirrors Python exactly
    process.exit(body ? 0 : 1);
  } else if (cmd === "get") {
    if (args.length < 2) {
      process.stderr.write("get requires a discussion number\n");
      process.exit(1);
    }
    const record = getDiscussion(parseInt(args[1]!, 10));
    console.log(JSON.stringify(record, null, 2));
  } else if (cmd === "list-open") {
    const records = listOpen();
    console.log(JSON.stringify(records, null, 2));
  } else if (cmd === "invalidate") {
    if (args.length < 2) {
      process.stderr.write("invalidate requires a discussion number\n");
      process.exit(1);
    }
    invalidate(parseInt(args[1]!, 10));
    console.log(`invalidated #${args[1]}`);
  } else if (cmd === "stats") {
    console.log(JSON.stringify(getStats(), null, 2));
  } else {
    process.stderr.write(`Unknown command: ${cmd}\n`);
    usage();
    process.exit(1);
  }
}
